grug/testing/src/client.rs

```diff
@@ -0,0 +1,313 @@
+};
+
+pub struct MockClient<DB = MemDb, VM = RustVm, PP = NaiveProposalPreparer, ID = NullIndexer>
+where
+    DB: Db,
+    VM: Vm,
+    PP: ProposalPreparer,
+    ID: Indexer,
+{
+    suite: Arc<Mutex<TestSuite<DB, VM, PP, ID>>>,
+    blocks: Arc<Mutex<BTreeMap<u64, (Block, BlockOutcome)>>>,
+    txs: Arc<Mutex<BTreeMap<Hash256, SearchTxOutcome>>>,
+    block_mode: BlockModeCache,
+}
+
+impl<DB, VM, PP, ID> MockClient<DB, VM, PP, ID>
+where
+    DB: Db + Send + Sync + 'static,
+    VM: Vm + Clone + Send + Sync + 'static,
+    PP: ProposalPreparer + Send + Sync + 'static,
+    ID: Indexer + Send + Sync + 'static,
+    AppError: From<DB::Error> + From<VM::Error> + From<PP::Error> + From<ID::Error>,
+{
+    pub fn new(suite: TestSuite<DB, VM, PP, ID>, block_mode: BlockCreation) -> Self {
+        let suite = Arc::new(Mutex::new(suite));
+        let blocks = Arc::new(Mutex::new(BTreeMap::new()));
+        let txs = Arc::new(Mutex::new(BTreeMap::new()));
+
+        let block_mode = match block_mode {
+            BlockCreation::Timed => {
+                let buffer = Arc::new(Mutex::new(vec![]));
+
+                let th_buffer = buffer.clone();
+                let th_suite = suite.clone();
+                let th_blocks = blocks.clone();
+                let th_txs = txs.clone();
+
+                thread::spawn(move || {
+                    let rt = Runtime::new().unwrap();
+
+                    rt.block_on(async move {
+                        loop {
+                            let sleep = th_suite.lock().await.block_time.into_nanos();
+                            tokio::time::sleep(Duration::from_nanos(sleep as u64)).await;
+
+                            let txs = {
+                                let mut guard = th_buffer.lock().await;
+                                std::mem::take(&mut *guard)
+                            };
+
+                            next_block(
+                                th_txs.lock().await.deref_mut(),
+                                th_blocks.lock().await.deref_mut(),
+                                th_suite.lock().await.deref_mut(),
+                                txs,
+                            )
+                            .await
+                            .unwrap();
+                        }
+                    })
+                });
+
+                BlockModeCache::Timed { buffer }
+            },
+            BlockCreation::OnBroadcast => BlockModeCache::OnBroadcast,
+        };
+
+        Self {
+            suite,
+            blocks,
+            txs,
+            block_mode,
+        }
+    }
+
+    pub async fn set_timestamp(&self, timestamp: Timestamp) {
+        self.suite.lock().await.block.timestamp = timestamp;
+    }
+
+    pub async fn set_block_time(&self, block_time: grug_types::Duration) {
+        self.suite.lock().await.block_time = block_time;
+    }
+}
+
+#[async_trait]
+impl<DB, VM, PP, ID> QueryClient for MockClient<DB, VM, PP, ID>
```

Comment:
When looking at the method bodies for those, I'm wondering about the use of having two traits: `QueryClientAsync` and `QueryClient` and get `query_app_async` and `query_app` (and same for other methods). I do understand the real implementation needs `async` tho, and having 2 different traits is probably not doable as for the code calling them.

Just throwing ideas here.